### PR TITLE
fix(llms): send reasoning_effort to every openai reasoning model

### PR DESCRIPTION
**Pull request by joaomdmoura**
## Related issue

None open for this. Found while measuring extraction latency on gpt-5: one run took 6.2s with `reasoning_effort` applied and 149.7s with it silently dropped.

## Summary

`reasoning_effort` never reached most OpenAI reasoning models on the chat-completions path. It was gated behind `is_o1_model = "o1" in model.lower()`, a literal substring test, so `gpt-5`, `gpt-5-mini`, `o3`, `o3-mini` and `o4-mini` all failed it and thought at the server default. The request still succeeded, so nothing surfaced.

The gate could not simply be widened: `is_o1_model` also drives `supports_function_calling`, `supports_stop_words` and the system-to-user message rewrite, so marking gpt-5 as an o1 model would report that it cannot call tools. A separate predicate decides it instead.

- **Shape, not a list.** `_supports_reasoning_effort` matches the o-series (`^o\d`) and GPT generation 5 onwards (`^gpt-(\d+)` >= 5), after stripping a provider prefix and the `ft:` fine-tune prefix, so a fine-tune is judged by its base model. `gpt-6` and `o5` already classify; `gpt-4o` and `gpt-4.1` parse to generation 4 and are excluded, so a non-reasoning model never pays a wasted round trip.
- **Compatible servers honour the explicit setting.** When the client points anywhere other than `api.openai.com` (explicit `base_url`/`api_base`, a `client_params` override as `str` or `httpx.URL`, `OPENAI_BASE_URL`/`OPENAI_API_BASE`, `custom_openai=True`, or a provider subclass such as OpenRouter, Ollama or vLLM), the model name is the server's namespace and says nothing about support, so a configured `reasoning_effort` is sent as is. This is what lets `gpt-oss-120b` on Groq, vLLM or OpenRouter, or a gateway deployment named `gpt-4o`, receive it.
- **Rejections are recovered once.** A 400 (or 422) that says the parameter is one the server does not know, in OpenAI's two shapes (`code="unsupported_parameter"`, "Unrecognized request argument") or the common wordings of compatible servers ("unknown field", "Extra inputs are not permitted"), is retried once without it. A 400 that merely names the parameter, a bad *value* for instance, is not: the model takes the parameter, and dropping it there would silently restore the original bug. The rejecting (endpoint, model) is remembered for the process once the retry has succeeded, so the wasted call is paid once and a retry that fails for an unrelated reason does not silence the setting, and the drop is logged as a warning because a configured setting is not being applied. The existing GPT-5.6 tools retry is untouched and still recovers by sending `"none"`.
- **`LLM.reasoning_effort` accepts `"minimal"` and `"xhigh"`**, the values the SDK accepts that the `Literal` omitted. Widening a `Literal` is non-breaking; the native provider already took them as untyped `str`.

`is_o1_model`, the Azure provider, and the litellm and Responses paths are unchanged.

## Verification

- [x] Tests added or updated for the changed behavior
- [x] Relevant tests and quality checks pass locally

`tests/llms/openai/test_reasoning_effort_forwarding.py` (88 tests) asserts the parameter on the built request across six models; that non-reasoning models never receive it; the shape predicate across supported, unsupported, fine-tuned and unreleased families; forwarding on compatible servers reached by flag, `base_url`, `client_params`, env and the `LLM` factory, and not for `base_url="https://api.openai.com/v1"`; the retry across sync, async, streaming and async streaming; recovery of a rejection worded outside the OpenAI shape (string body, FastAPI 422); that an unsupported *value*, a pydantic enum complaint, or an ambiguous 400 naming the parameter still raises and is not remembered; that a gateway rejection, whether the gateway is set by `base_url` or by an `httpx.URL` in `client_params`, does not silence the same model on OpenAI; that a rejection is remembered only once the retry succeeds, on the sync and async paths; that a rejecting model is not sent the parameter again; and that `is_o1_model`'s other three behaviours are unchanged.

Locally: `pytest tests/llms tests/test_llm.py` 880 passed, 23 skipped; `ruff check lib/`, `ruff format --check lib/` and `mypy` clean. Measured end to end on gpt-5 before the fix: 6.2s with the setting applied against 149.7s with it dropped.

Worth knowing for review: `tests/test_llm.py` already had three `test_o3_mini_reasoning_effort_*` tests. They pass either way, since they assert `"Paris" in result` and VCR's default matcher ignores the request body; their cassettes were recorded in the litellm era. That is why this went unnoticed.

## Additional context

- Behaviour change on compatible servers: a configured `reasoning_effort` is now sent where before it was silently dropped unless the model name contained "o1". A server that says the parameter is unknown is recovered by the retry; one that rejects it in words the matcher does not recognise now surfaces that 400, which is the caller's configuration to fix.
- Known over-matches on OpenAI's own API (`gpt-5-chat-latest`, `o1-mini`, `o1-preview` do not take the parameter) pay one rejected call per process, then are remembered.
- `lib/crewai-tools/tool.specs.json` embeds the `LLM` schema with the old four-value enum. The regenerating workflow only triggers on crewai-tools changes, so run `generate-tool-specs` via workflow_dispatch after merge.
- No docs change: `docs/edge/en/concepts/llms.mdx` shows `reasoning_effort` only in usage examples and never enumerates values or supported models.
- Follow-ups: the three `o3-mini` cassettes want re-recording with a body matcher so they assert something; the Azure chat-completions path never sends `reasoning_effort` at all (pre-existing, separate provider).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Changes core OpenAI completion request shaping and retry behavior for all native OpenAI chat paths; misclassification could alter latency/cost or surface new 400s on compatible gateways, though retries and per-process learning limit blast radius.
> 
> **Overview**
> Fixes a bug where **`reasoning_effort` was only attached to chat-completions requests when the model name contained `"o1"`**, so `gpt-5`, `o3`, `o4-mini`, and similar models ignored the setting while calls still succeeded.
> 
> **Forwarding:** A separate **`_supports_reasoning_effort`** check (o-series and GPT generation ≥ 5, with fine-tune prefix handling) replaces that gate without widening **`is_o1_model`**, which still controls tools, stop words, and message rewriting. On **OpenAI-compatible endpoints** (non-`api.openai.com`), an explicit `reasoning_effort` is sent regardless of model name.
> 
> **Resilience:** If the API returns 400/422 indicating the parameter is unknown (OpenAI shapes or common gateway wordings), the client **retries once without `reasoning_effort`**, remembers the **(endpoint, model)** pair only after a successful retry, and logs a warning. Invalid *values* are not retried or silently dropped.
> 
> **API surface:** `LLM.reasoning_effort` now allows **`minimal`** and **`xhigh`**. A large new test module covers forwarding, retries, compatible servers, and edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c4eb01a4b6e65652fa5791109255b891e89166cf. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: lib/crewai/src/crewai/llms/providers/openai/completion.py [hunk 1]
```diff
@@ -70,6 +71,27 @@
...
+def _supports_reasoning_effort(model: str) -> bool:
+    """Whether the model accepts `reasoning_effort` on /v1/chat/completions."""
+    name = model.rsplit("/", 1)[-1].lower()
+    if _O_SERIES_MODEL.match(name):
+        return True
+    generation = _GPT_GENERATION.match(name)
+    return (
+        generation is not None
+        and int(generation.group(1)) >= _MIN_REASONING_GPT_GENERATION
+    )
```
[the remark this answers is not in the document]
Comment (joaomdmoura) [reply lib/crewai/src/crewai/llms/providers/openai/completion.py:218]: Addressed in 821f94862. Fine-tunes: the `ft:` prefix is stripped and the base model decides, so `ft:o4-mini-...` matches and `ft:gpt-4o-mini-...` does not. Compatible servers, which is where `gpt-oss-*` and the inherited providers actually live: the shape match now applies only when the client points at `api.openai.com`. Anywhere else (explicit `base_url`/`api_base`, `OPENAI_BASE_URL`, `custom_openai=True`, or a provider subclass) a configured `reasoning_effort` is sent as is, and a 400 that names the parameter in the server's own words is recovered by one retry without it, with the rejecting (endpoint, model) remembered for the process. Complaints about the value still surface. Tests cover `gpt-oss-120b` through the `LLM` factory with a `base_url`, the OpenRouter subclass, env-configured URLs, and string-body and FastAPI-422 rejections. The one slice not covered is `gpt-oss-*` named directly against api.openai.com with no base URL, which OpenAI does not serve.

File: lib/crewai/src/crewai/llms/providers/openai/completion.py [hunk 2]
```diff
@@ -1829,8 +1914,11 @@ def _prepare_completion_params(
         if self.top_logprobs is not None:
             params["top_logprobs"] = self.top_logprobs
 
-        # Handle o1 model specific parameters
-        if self.is_o1_model and self.reasoning_effort:
+        # Not gated on `is_o1_model`: that is a literal "o1" substring test, so
+        # gpt-5, o3 and o4-mini failed it and silently thought at the server
+        # default. It also drives tool support and message rewriting, so it
+        # cannot be widened to mean "is a reasoning model".
+        if self.reasoning_effort and _supports_reasoning_effort(self.model):
```
[the remark this answers is not in the document]
Comment (joaomdmoura) [reply lib/crewai/src/crewai/llms/providers/openai/completion.py]: Addressed in 821f94862. `custom_openai=True` now marks the server compatible regardless of the deployment name, so `gpt-4o` and `gpt-4.1` deployments receive the configured value. The same holds when the server is reached by `base_url`/`api_base` or the env variables without the flag, since the `LLM` factory only sets `custom_openai` on some of those paths. If the endpoint rejects the parameter, one retry without it recovers the call and the (endpoint, model) is remembered, so a rejection on a gateway cannot silence the same model name on OpenAI itself.

File: lib/crewai/src/crewai/llms/providers/openai/completion.py [hunk 3]
```diff
@@ -1862,6 +1912,110 @@ def _rejects_reasoning_effort_with_tools(error: BaseException) -> bool:
...
+        return self._on_compatible_server() or _supports_reasoning_effort(self.model)
+
+    def _effective_base_url(self) -> str | None:
+        """The base URL the client is built with; None means OpenAI's own API."""
+        return (
+            self.base_url
+            or self.api_base
+            or os.getenv("OPENAI_BASE_URL")
+            or os.getenv("OPENAI_API_BASE")
+        )
```
[the remark this answers is not in the document]
Comment (joaomdmoura) [reply lib/crewai/src/crewai/llms/providers/openai/completion.py:1957]: Fixed in 1dfd03612: `_effective_base_url` now applies the same precedence as `_get_client_params`, so a `client_params["base_url"]` override marks the server compatible and keys the remembered rejection under that endpoint. Test added.

File: lib/crewai/src/crewai/llms/providers/openai/completion.py [hunk 4]
```diff
@@ -1862,6 +1912,110 @@ def _rejects_reasoning_effort_with_tools(error: BaseException) -> bool:
...
+            return False
+        if source.get("code") == "unsupported_parameter":
+            return True
+        if "unrecognized request argument" in message:
+            return True
+        if not self._on_compatible_server():
+            return False
+        if source.get("code") in {"unsupported_value", "invalid_value"}:
+            return False
+        return not any(hint in message for hint in _REASONING_EFFORT_VALUE_COMPLAINTS)
```
[the remark this answers is not in the document]
Comment (joaomdmoura) [reply lib/crewai/src/crewai/llms/providers/openai/completion.py]: Agreed, and fixed in 1dfd03612. The matcher now requires positive evidence that the field is unknown: OpenAI's two shapes plus the common compatible-server wordings ("unknown field", "unknown parameter", "unexpected field", "extra inputs", "not permitted", "additional properties"). Anything else that names the parameter, including the pydantic `Input should be 'low', 'medium' or 'high'` case you cite, surfaces and is not remembered. The negative word list is gone. Tests cover the enum message and an ambiguous 400.

File: lib/crewai/src/crewai/llms/providers/openai/completion.py [hunk 5]
```diff
@@ -1862,6 +1924,111 @@ def _rejects_reasoning_effort_with_tools(error: BaseException) -> bool:
...
+            return False
+        return self._on_compatible_server() or _supports_reasoning_effort(self.model)
+
+    def _effective_base_url(self) -> str | None:
+        """The base URL the client is built with; None means OpenAI's own API.
+
+        Same precedence as `_get_client_params`, where `client_params` win.
+        """
+        override = (self.client_params or {}).get("base_url")
+        if isinstance(override, str) and override:
```
Comment (joaomdmoura) [inline lib/crewai/src/crewai/llms/providers/openai/completion.py]: OpenAI 2.41.0 accepts `str | httpx.URL` for `base_url`, and `_get_client_params()` forwards either. Can we normalize both here so endpoint detection and the rejection key match the endpoint the client calls? Otherwise a compatible deployment named `gpt-4o` silently loses its configured `reasoning_effort`.

File: lib/crewai/src/crewai/llms/providers/openai/completion.py [hunk 6]
```diff
@@ -649,6 +699,12 @@ def dispatch(params: dict[str, Any]) -> str | Any:
                     )
                     return dispatch(retry_params)
 
+            if self._rejects_reasoning_effort_as_unsupported(cause):
+                retry_params = self._without_reasoning_effort(completion_params)
+                if retry_params is not None:
+                    self._remember_no_reasoning_effort_model()
```
Comment (joaomdmoura) [inline lib/crewai/src/crewai/llms/providers/openai/completion.py:706]: Can we remember the rejection only after this retry succeeds? If the fallback fails for an unrelated transient reason, later calls silently stop sending the configured effort even though the fallback was never established.

File: lib/crewai/tests/llms/openai/test_reasoning_effort_forwarding.py [hunk 7]
```diff
@@ -0,0 +1,618 @@
...
+                    ]
+                },
+            ),
+            _status_error(
+                BadRequestError, 400, {"error": "reasoning_effort: something odd"}
+            ),
+        ],
+        ids=["pydantic-enum", "ambiguous"],
+    )
+    def test_without_evidence_the_parameter_is_unknown_the_error_surfaces(
```
Comment (joaomdmoura) [inline lib/crewai/tests/llms/openai/test_reasoning_effort_forwarding.py:632]: Good adversarial coverage. Requiring positive evidence that the field is unknown prevents invalid configured values from silently becoming server defaults.

**Review by joaomdmoura — COMMENTED** [review]
This fixes the silent `reasoning_effort` omission while keeping the old `is_o1_model` behavior intact, and the positive error matcher avoids hiding invalid values.

I’m requesting changes for two cases in the inline comments: `httpx.URL` overrides can disagree with the endpoint the client calls, and rejection memory is committed before the fallback succeeds. I did not run the tests locally or reproduce the latency measurement. CI is green.

I'd want someone else to sign off here too, mainly because `LLM.reasoning_effort` expands a public configuration contract with two accepted values. (I can't formally request changes on my own PR, so this goes in as a comment.).

**Comment by Vidit-Ostwal** [comment]
@code-rabbit full review?

**Review by Vidit-Ostwal — COMMENTED** [review]
`_effective_base_url()` only recognizes string `client_params["base_url"]` values, but `_get_client_params()` forwards `httpx.URL` values and the OpenAI SDK accepts them. This means a compatible host configured with `client_params={"base_url": httpx.URL("https://gateway.example/v1")}` is treated as OpenAI for capability detection: `reasoning_effort` is dropped for gateway aliases such as `gpt-4o`.

For model names that still match the local predicate (for example `gpt-5`), an unsupported-parameter response is also cached under `(api.openai.com, model)`, potentially suppressing the setting on a later genuine OpenAI request.

Please normalize URL-like overrides when resolving the effective base URL, and add regressions for:

1. `client_params["base_url"]` as `httpx.URL` forwarding the setting to a compatible server.
2. A rejection from that URL-valued gateway remaining isolated from OpenAI’s endpoint cache.

File: lib/crewai/src/crewai/llms/providers/openai/completion.py [hunk 8]
```diff
@@ -1862,6 +1924,111 @@ def _rejects_reasoning_effort_with_tools(error: BaseException) -> bool:
...
+        return self._on_compatible_server() or _supports_reasoning_effort(self.model)
+
+    def _effective_base_url(self) -> str | None:
+        """The base URL the client is built with; None means OpenAI's own API.
+
+        Same precedence as `_get_client_params`, where `client_params` win.
+        """
+        override = (self.client_params or {}).get("base_url")
+        if isinstance(override, str) and override:
+            return override
```
[the remark this answers is not in the document]
Comment (joaomdmoura) [reply lib/crewai/src/crewai/llms/providers/openai/completion.py]: Fixed in 991bbd7bf. `_effective_base_url` now accepts the same `str | httpx.URL` the SDK does and normalises both to the string the client calls, so a `client_params={"base_url": httpx.URL(...)}` gateway is detected as a compatible server and its rejection is keyed under that host, not `api.openai.com`.

Tests: `test_an_httpx_url_override_is_the_endpoint_the_client_calls` (a `gpt-4o` behind a URL-valued gateway receives the setting, and the memory key is the gateway), and `test_a_gateway_rejection_does_not_silence_the_model_on_openai` is now parametrised over `base_url=` and `client_params={"base_url": httpx.URL(...)}`, so the isolation from OpenAI is pinned for both forms.

**Comment by joaomdmoura** [comment]
@Vidit-Ostwal thanks, both points are in 991bbd7bf.

`_effective_base_url` now normalises `str | httpx.URL` overrides in `client_params` to the string the client calls, so endpoint detection and the rejection key agree with the endpoint the SDK actually uses. The same commit also moves the rejection memory to after the retry returns: a retry that fails for an unrelated reason no longer silences the configured effort for the rest of the process.

The two regressions you asked for:

1. `test_an_httpx_url_override_is_the_endpoint_the_client_calls`: a `gpt-4o` behind `client_params={"base_url": httpx.URL(gateway)}` receives the setting, and its memory key is `(gateway, "gpt-4o")`.
2. `test_a_gateway_rejection_does_not_silence_the_model_on_openai[httpx.URL]`: a `gpt-5` rejected by a URL-valued gateway is remembered under the gateway, and a plain OpenAI `gpt-5` still sends the setting afterwards. The existing `base_url=` case is the other parameter of the same test.

Locally: the forwarding file is 88 passed; `pytest lib/crewai/tests/llms lib/crewai/tests/test_llm.py` is 880 passed, 23 skipped; mypy is clean on `completion.py`.

File: lib/crewai/src/crewai/llms/providers/openai/completion.py (hunk 5 above)
Comment (joaomdmoura) [reply lib/crewai/src/crewai/llms/providers/openai/completion.py]: Fixed in 991bbd7bf: `_effective_base_url` normalises `str | httpx.URL` to the string the client calls. Regressions: `test_an_httpx_url_override_is_the_endpoint_the_client_calls`, and the `httpx.URL` parameter of `test_a_gateway_rejection_does_not_silence_the_model_on_openai`.

File: lib/crewai/src/crewai/llms/providers/openai/completion.py (hunk 6 above)
Comment (joaomdmoura) [reply lib/crewai/src/crewai/llms/providers/openai/completion.py:706]: Fixed in 991bbd7bf: the (endpoint, model) is added to the learned set only after `dispatch(retry_params)` returns, on both the sync and async paths; the streaming handlers consume the stream before returning, so this covers all four. Tests: `test_a_rejection_is_remembered_only_once_the_retry_succeeds` and its async twin, which fail the retry with an `APIConnectionError`, assert the set is still empty, then show the effort is sent again on the next call and dropped for good once a retry succeeds.

**Review by Vidit-Ostwal — COMMENTED** [review]
LGTM

**Review by Vidit-Ostwal — APPROVED** [review]
